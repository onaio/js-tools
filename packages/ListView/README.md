# ListView

Simple component for listing items in a table.

## How it works

## Installation & Usage

### Examples

## Props
