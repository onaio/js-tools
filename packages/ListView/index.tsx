import ElementMap from '@onaio/element-map';
import React from 'react';

/**
 * TODO
 * 1. custom class names for table, tbody, and thead
 * 2. entire thead as optional
 * 3. Future PR: custom renderHeaders and renderRows supplied as props
 */

/** Docstring goes here */
export interface ListViewProps {
  data: React.ReactNode[][];
  headerItems: any;
  tableClass: any;
  tbodyClass: any;
  theaderClass: any;
}

/** ListView */
const ListView: React.ElementType = (props: ListViewProps) => {
  const { data, headerItems, tableClass, tbodyClass, theaderClass } = props;

  /** changeme */
  function renderHeaders(items: any[], thClass: any) {
    if (items) {
      return (
        <th className={thClass ? thClass : null}>
          <tr>
            <ElementMap items={items} HTMLTag="th" />
          </tr>
        </th>
      );
    } else {
      return null;
    }
  }

  /** changeme */
  function renderRows(rowData: React.ReactNode[][], tbClass: any) {
    const result = rowData.map((item, itemKey) => (
      <tr key={itemKey}>
        <ElementMap items={[item]} HTMLTag="td" />
      </tr>
    ));
    return <tbody className={tbClass ? tbClass : null}>{result}</tbody>;
  }

  const headers = renderHeaders(headerItems, theaderClass);
  const rows = renderRows(data, tbodyClass);

  return (
    <table className={tableClass ? tableClass : null}>
      {headers}
      {rows}
    </table>
  );
};

export default ListView;
