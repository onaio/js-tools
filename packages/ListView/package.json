{
  "name": "@onaio/list-view",
  "version": "0.0.1",
  "description": "list-view",
  "main": "index.js",
  "repository": "https://github.com/onaio/js-tools",
  "author": "Ona Engineering",
  "license": "Apache-2.0"
}
