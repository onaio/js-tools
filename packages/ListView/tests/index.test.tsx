import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import ListView, { ListViewProps } from '../';

describe('ListView', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    const props = {
      data: [['Data 1'], ['Data 2'], ['Data 3']],
      headerItems: ['Columns'],
      tableClass: 'table-striped',
      tbodyClass: 'table-active',
      theaderClass: 'thead-dark'
    };
    shallow(<ListView {...props} />);
  });

  it('renders correctly', () => {
    const props = {
      data: [['Data 1'], ['Data 2'], ['Data 3']],
      headerItems: ['Columns'],
      tableClass: 'table-striped',
      tbodyClass: 'table-active',
      theaderClass: 'thead-dark'
    };
    const wrapper = mount(<ListView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly with different types of data', () => {
    const props: ListViewProps = {
      data: [['Data 1'], [999], ['Data 3']],
      headerItems: ['Columns'],
      tableClass: 'table-striped',
      tbodyClass: 'table-active',
      theaderClass: 'thead-dark'
    };
    const wrapper = mount(<ListView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders correctly with rows containing components', () => {
    const SomeComponent: React.ElementType = props => {
      return (
        <div>
          <span>{props.value}</span>
        </div>
      );
    };
    const tableProps: ListViewProps = {
      data: [
        ['Data 1'],
        [<div key="123">Data 2</div>],
        [<SomeComponent key="456" value="Data 3" />]
      ],
      headerItems: ['Columns'],
      tableClass: 'table-striped',
      tbodyClass: 'table-active',
      theaderClass: 'thead-dark'
    };
    const wrapper = mount(<ListView {...tableProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
  it('renders correctly without th, tbclass, tbodyclass and theaderclass', () => {
    const props: ListViewProps = {
      data: [['Data 1'], [999], ['Data 3']]
    };
    const wrapper = mount(<ListView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
