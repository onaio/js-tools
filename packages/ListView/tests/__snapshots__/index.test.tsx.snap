// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListView renders correctly 1`] = `
<ListView
  data={
    Array [
      Array [
        "Ed",
        6,
        "Taller",
      ],
      Array [
        "Edd",
        12,
        "Tallest",
      ],
      Array [
        "Eddie",
        17,
        "Tall",
      ],
    ]
  }
  headerItems={
    Array [
      "Name",
      "Age",
      "Height",
    ]
  }
  renderHeaders={[Function]}
  tableClass="table-striped"
  tbodyClass="table-active"
  theadClass="thead-dark"
>
  <table
    className="table-striped"
  >
    <thead
      className="thead-dark"
    >
      <tr>
        <ElementMap
          HTMLTag="th"
          className="element-map"
          items={
            Array [
              "Name",
              "Age",
              "Height",
            ]
          }
        >
          <th
            className="element-map"
            key="0"
          >
            Name
          </th>
          <th
            className="element-map"
            key="1"
          >
            Age
          </th>
          <th
            className="element-map"
            key="2"
          >
            Height
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="table-active"
    >
      <tr
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Ed",
              6,
              "Taller",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Ed
          </td>
          <td
            className="element-map"
            key="1"
          >
            6
          </td>
          <td
            className="element-map"
            key="2"
          >
            Taller
          </td>
        </ElementMap>
      </tr>
      <tr
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Edd",
              12,
              "Tallest",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Edd
          </td>
          <td
            className="element-map"
            key="1"
          >
            12
          </td>
          <td
            className="element-map"
            key="2"
          >
            Tallest
          </td>
        </ElementMap>
      </tr>
      <tr
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Eddie",
              17,
              "Tall",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Eddie
          </td>
          <td
            className="element-map"
            key="1"
          >
            17
          </td>
          <td
            className="element-map"
            key="2"
          >
            Tall
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly with custom renderHeaders prop 1`] = `
<ListView
  data={
    Array [
      Array [
        "Ed",
        6,
      ],
      Array [
        "Edd",
        12,
      ],
      Array [
        "Eddie",
        17,
      ],
    ]
  }
  headerItems={
    Array [
      "Name",
      "Age",
    ]
  }
  renderHeaders={[Function]}
  tableClass="listview"
  tbodyClass="listview-tbody"
  theadClass="listview-thead"
>
  <table
    className="listview"
  >
    <thead
      className="listview-thead"
    >
      <tr>
        <th
          colSpan={2}
        >
          Top Header
        </th>
      </tr>
      <tr>
        <ElementMap
          HTMLTag="th"
          className="element-map"
          items={
            Array [
              "Name",
              "Age",
            ]
          }
        >
          <th
            className="element-map"
            key="0"
          >
            Name
          </th>
          <th
            className="element-map"
            key="1"
          >
            Age
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="listview-tbody"
    >
      <tr
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Ed",
              6,
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Ed
          </td>
          <td
            className="element-map"
            key="1"
          >
            6
          </td>
        </ElementMap>
      </tr>
      <tr
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Edd",
              12,
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Edd
          </td>
          <td
            className="element-map"
            key="1"
          >
            12
          </td>
        </ElementMap>
      </tr>
      <tr
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Eddie",
              17,
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Eddie
          </td>
          <td
            className="element-map"
            key="1"
          >
            17
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly with different types of data 1`] = `
<ListView
  data={
    Array [
      Array [
        "Data 1",
      ],
      Array [
        999,
      ],
      Array [
        "Data 3",
      ],
    ]
  }
  headerItems={
    Array [
      "Columns",
    ]
  }
  renderHeaders={[Function]}
  tableClass="table-striped"
  tbodyClass="table-active"
  theadClass="thead-dark"
>
  <table
    className="table-striped"
  >
    <thead
      className="thead-dark"
    >
      <tr>
        <ElementMap
          HTMLTag="th"
          className="element-map"
          items={
            Array [
              "Columns",
            ]
          }
        >
          <th
            className="element-map"
            key="0"
          >
            Columns
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="table-active"
    >
      <tr
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Data 1",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Data 1
          </td>
        </ElementMap>
      </tr>
      <tr
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              999,
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            999
          </td>
        </ElementMap>
      </tr>
      <tr
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Data 3",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Data 3
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly with no renderHeaders prop 1`] = `
<ListView
  data={
    Array [
      Array [
        "Ed",
      ],
      Array [
        "Edd",
      ],
      Array [
        "Eddie",
      ],
    ]
  }
  headerItems={
    Array [
      "Name",
    ]
  }
  renderHeaders={[Function]}
  tableClass="listview"
  tbodyClass="listview-tbody"
  theadClass="listview-thead"
>
  <table
    className="listview"
  >
    <thead
      className="listview-thead"
    >
      <tr>
        <ElementMap
          HTMLTag="th"
          className="element-map"
          items={
            Array [
              "Name",
            ]
          }
        >
          <th
            className="element-map"
            key="0"
          >
            Name
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="listview-tbody"
    >
      <tr
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Ed",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Ed
          </td>
        </ElementMap>
      </tr>
      <tr
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Edd",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Edd
          </td>
        </ElementMap>
      </tr>
      <tr
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Eddie",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Eddie
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly with rows containing components 1`] = `
<ListView
  data={
    Array [
      Array [
        "Data 1",
      ],
      Array [
        <div>
          Data 2
        </div>,
      ],
      Array [
        <SomeComponent
          value="Data 3"
        />,
      ],
    ]
  }
  headerItems={
    Array [
      "Columns",
    ]
  }
  renderHeaders={[Function]}
  tableClass="table-striped"
  tbodyClass="table-active"
  theadClass="thead-dark"
>
  <table
    className="table-striped"
  >
    <thead
      className="thead-dark"
    >
      <tr>
        <ElementMap
          HTMLTag="th"
          className="element-map"
          items={
            Array [
              "Columns",
            ]
          }
        >
          <th
            className="element-map"
            key="0"
          >
            Columns
          </th>
        </ElementMap>
      </tr>
    </thead>
    <tbody
      className="table-active"
    >
      <tr
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Data 1",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Data 1
          </td>
        </ElementMap>
      </tr>
      <tr
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              <div>
                Data 2
              </div>,
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            <div
              key="123"
            >
              Data 2
            </div>
          </td>
        </ElementMap>
      </tr>
      <tr
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              <SomeComponent
                value="Data 3"
              />,
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            <SomeComponent
              key="456"
              value="Data 3"
            >
              <div>
                <span>
                  Data 3
                </span>
              </div>
            </SomeComponent>
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;

exports[`ListView renders correctly without th, tbclass, tbodyclass and theadClass 1`] = `
<ListView
  data={
    Array [
      Array [
        "Data 1",
      ],
      Array [
        999,
      ],
      Array [
        "Data 3",
      ],
    ]
  }
  renderHeaders={[Function]}
  tableClass="listview"
  tbodyClass="listview-tbody"
  theadClass="listview-thead"
>
  <table
    className="listview"
  >
    <tbody
      className="listview-tbody"
    >
      <tr
        key="0"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Data 1",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Data 1
          </td>
        </ElementMap>
      </tr>
      <tr
        key="1"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              999,
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            999
          </td>
        </ElementMap>
      </tr>
      <tr
        key="2"
      >
        <ElementMap
          HTMLTag="td"
          className="element-map"
          items={
            Array [
              "Data 3",
            ]
          }
        >
          <td
            className="element-map"
            key="0"
          >
            Data 3
          </td>
        </ElementMap>
      </tr>
    </tbody>
  </table>
</ListView>
`;
